server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -322,51 +322,55 @@ private static Map<String, Map<String, FieldCapabilities>> buildResponseMap(
         Map<String, Map<String, FieldCapabilities.Builder>> responseMapBuilder,
         boolean includeUnmapped
     ) {
-        Map<String, Map<String, FieldCapabilities>> responseMap = new HashMap<>();
+        Map<String, Map<String, FieldCapabilities>> responseMap = Maps.newMapWithExpectedSize(responseMapBuilder.size());
+        final var indices = indexResponsesMap.keySet();
         for (Map.Entry<String, Map<String, FieldCapabilities.Builder>> entry : responseMapBuilder.entrySet()) {
-            Map<String, FieldCapabilities.Builder> typeMapBuilder = entry.getValue();
+            var typeMapBuilder = entry.getValue().entrySet();
 
-            Optional<Function<Boolean, FieldCapabilities>> unmapped = Optional.empty();
+            Function<Boolean, FieldCapabilities> unmapped = null;
             if (includeUnmapped) {
                 // do this directly, rather than using the builder, to save creating a whole lot of objects we don't need
                 unmapped = getUnmappedFields(
-                    indexResponsesMap.keySet(),
+                    indices,
                     entry.getKey(),
-                    typeMapBuilder.values().stream().flatMap(FieldCapabilities.Builder::getIndices).collect(Collectors.toSet())
+                    typeMapBuilder.stream().flatMap(t -> t.getValue().getIndices()).collect(Collectors.toSet())
                 );
             }
 
-            boolean multiTypes = typeMapBuilder.size() + unmapped.map(f -> 1).orElse(0) > 1;
-            responseMap.put(
-                entry.getKey(),
-                Collections.unmodifiableMap(
-                    Stream.concat(
-                        typeMapBuilder.entrySet()
-                            .stream()
-                            .map(e -> Map.<String, Function<Boolean, FieldCapabilities>>entry(e.getKey(), e.getValue()::build)),
-                        unmapped.stream().map(f -> Map.entry("unmapped", f))
-                    ).collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().apply(multiTypes)))
-                )
-            );
+            final int resSize = typeMapBuilder.size() + (unmapped == null ? 0 : 1);
+            boolean multiTypes = resSize > 1;
+            final Map<String, FieldCapabilities> res = Maps.newHashMapWithExpectedSize(resSize);
+            for (Map.Entry<String, FieldCapabilities.Builder> e : typeMapBuilder) {
```

Comment:
Building to a pre-sized map outright without all the endless intermediary objects should save quite a bit of heap here.